### PR TITLE
usnic: Explicitly pad the fi_usnic_info struct.

### LINKED ISSUE #2454: usnic: getinfo extension in v1.3.0 and above is not compatible with older releases
PR #1866 added version 2 of the `fi_usnic_info` extension that improved upon version 1 by providing extra information about Cisco usNIC hardware. Although versioning precautions were taken, there were mistakes in the implementation of the original extension as well as the updates made in version 1.3.

1. The structure were not declared using the `__attribute__ ((packed))` extension. This means that the size and alignment of the structures were variable due to structure packing imposed by the compiler.
1. Several of the new fields added in the 1.3 update were 64-bit pointers. The structure previously did not contain any 64-bit members so the structure had a 4-byte alignment. Once this change was introduced, the union became aligned on an 8-byte boundary and shifted the start offset of the v1 and v2 fields.

These two mistakes can lead to invalid data when the runtime and compile time versions of Libfabric are different.

Release 1.4 will explicitly pad the structure to conform to the alignments in the 1.3 release. If this bug is encountered, then it is recommended that the user update to a release containing the version 2 extension or retrieve a patched version of an older release.

### PR DESCRIPTION
**Pull request by bturrubiates**
The introduction of version 2 of the fi_usnic_info extension caused a
compatibility issue with older versions of the extension. This commit
adds padding to the structure to explicitly maintain the alignment
requirements for the extension in the 1.3 release. Also, mark the outer
structure as having the packed attribute.

Relevant documents have been updated to document the issue and recommend
possible solutions.

Fixes #2454 

Signed-off-by: Ben Turrubiates <bturrubi@cisco.com>

### CODE REVIEW

File: man/fi_usnic.7.md [hunk 1]
```diff
@@ -98,17 +98,30 @@ request a version rather than using the header provided
 `FI_EXT_USNIC_INFO_VERSION`. Although there is a version 1 of the extension,
 its use is discouraged, and it may not be available in future releases.
 
+## Fabric Extension: getinfo: compatibility issues
```
Comment (jsquyres) [inline man/fi_usnic.7.md]: Minor nit: Make this a `###` (i.e., a sub of the larger `Fabric Extension: getinfo` section).  And just call it "Compatibility issues".

File: man/fi_usnic.7.md [hunk 2]
```diff
@@ -98,17 +98,30 @@ request a version rather than using the header provided
 `FI_EXT_USNIC_INFO_VERSION`. Although there is a version 1 of the extension,
```
Comment (jsquyres) [inline man/fi_usnic.7.md]: On line 91, do we need to say "FI_USNIC_FABRIC_OPS_1 or FI_USNIC_FABRIC_OPS_2 ..."?

File: man/fi_usnic.7.md [hunk 3]
```diff
@@ -98,17 +98,30 @@ request a version rather than using the header provided
...
 its use is discouraged, and it may not be available in future releases.
 
+## Fabric Extension: getinfo: compatibility issues
+
+The addition of version 2 of the extension caused an alignmment issue that
+could lead to invalid data in the v1 portion of the structure. The issue
+manifests when compiling and running with different releases of Libfabric in
+which only one release contains version 2 of the extension.
```
Comment (jsquyres) [inline man/fi_usnic.7.md]: I think it's ok to be specific / clear here:

> The issue manifests when compiling an application against Libfabric v1.1.x or v1.2.x and then running with `libfabric.so` that is v1.3.x or higher.

File: prov/usnic/src/fi_ext_usnic.h [hunk 4]
```diff
@@ -104,13 +104,22 @@ struct fi_usnic_info_v2 {
 	struct fi_usnic_cap	**ui_caps;
 };
 
+/* In API version 1.2 and below, the v1 structure did not contain any 64-bit
+ * data types and therefore had a 4-byte alignment. Once v2 of the extension API
+ * was introduced in version 1.3, the extra pointers mandated an 8-byte
+ * alignment thus changing the offset of the v1 structure. This means that
+ * the post-1.3 getinfo extension API is not compatible with the previous
+ * versions (unless patched). Make the alignment explicit and consistent by
```
Comment (jsquyres) [inline prov/usnic/src/fi_ext_usnic.h]: Replace the "This means that..." sentence with similar language from the man page:

> This means that the alignment difference manifests when an application using v1 of the extension is compiled Libfabric v1.1.x or v1.2.x, but then runs with `libfabric.so` that is v1.3.x or higher.

**Comment by jsquyres** [comment]
@bturrubiates Do you know if adding the packed attribute changes the packing of the v2 struct?  Or does it only affect the packing of the outer struct?

**Comment by bturrubiates** [comment]
> Do you know if adding the packed attribute changes the packing of the v2 struct? Or does it only affect the packing of the outer struct?

It should only affect the outer structure. I've checked using [pahole](https://linux.die.net/man/1/pahole), and that seems to be the case. Whether, that's _always_ the case though...might be a better question for @goodell. AFAIK, the packed attribute _only_ affects the struct that is marked. It doesn't really matter here at the moment since there aren't any holes in the `fi_usnic_info` struct after adding the packing. This was just for future changes...

File: man/fi_usnic.7.md (hunk 2 above)
Comment (bturrubiates) [inline man/fi_usnic.7.md]: There is no `FI_USNIC_FABRIC_OPS_2`. It's correct as-is.

File: prov/usnic/src/fi_ext_usnic.h [hunk 5]
```diff
@@ -102,15 +110,25 @@ struct fi_usnic_info_v2 {
...
-};
-
+} __attribute__((packed));
+
+/* In API version 1.2 and below, the v1 structure did not contain any 64-bit
+ * data types and therefore had a 4-byte alignment. Once v2 of the extension API
+ * was introduced in version 1.3, the extra pointers mandated an 8-byte
+ * alignment thus changing the offset of the v1 structure. This means that the
+ * alignment difference manifests when an application using v1 of the extension
+ * is compiled Libfabric v1.1.x or v1.2.x, but then runs with libfabric.so that
```
Comment (goodell) [inline prov/usnic/src/fi_ext_usnic.h]: missing "with" here

File: man/fi_usnic.7.md (hunk 2 above)
Comment (jsquyres) [inline man/fi_usnic.7.md]: Right -- got it...

**Pull request merged**